plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
 	return true, nil
 }
 
+type MaxEBSVolumeCountChecker struct {
+	maxVolumes uint
+	pvInfo     PersistentVolumeInfo
+	pvcInfo    PersistentVolumeClaimInfo
+}
+
+// MaxEBSVolumeCountPredicate evaluates whether a pod can fit based on the number of AWS ElasticBlockStorage
+// volumes it requests, and those that are already present.  The maximum number is configurable, although
+// Amazon reccomends that only 40 volumes be mounted, including the system root volume.
+//
+// The predicate looks for both EBS volumes used directly, as well as PVC volumes that are backed by EBS volumes,
+// counts the number of unique volumes, and rejects the new pod if it would place the total count over the maximum.
+func NewMaxEBSVolumeCountPredicate(maxVolumes uint, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
+	c := &MaxEBSVolumeCountChecker{
+		maxVolumes: maxVolumes,
+		pvInfo:     pvInfo,
+		pvcInfo:    pvcInfo,
+	}
+
+	return c.predicate
+}
+
+func (c *MaxEBSVolumeCountChecker) filterEBSVolumes(volumes []api.Volume, namespace string, ebsVolumes map[string]bool) error {
+	for _, vol := range volumes {
+		if vol.AWSElasticBlockStore != nil {
+			ebsVolumes[vol.AWSElasticBlockStore.VolumeID] = true
+		} else if vol.PersistentVolumeClaim != nil {
+			pvcName := vol.PersistentVolumeClaim.ClaimName
+			if pvcName == "" {
+				return fmt.Errorf("PersistentVolumeClaim had no name: %q", pvcName)
+			}
+			pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
+			if err != nil {
+				return err
+			}
+
+			if pvc == nil {
+				return fmt.Errorf("PersistentVolumeClaim was not found: %q", pvcName)
+			}
+
+			pvName := pvc.Spec.VolumeName
+			if pvName == "" {
+				return fmt.Errorf("PersistentVolumeClaim is not bound: %q", pvcName)
+			}
+
+			pv, err := c.pvInfo.GetPersistentVolumeInfo(pvName)
+			if err != nil {
+				return err
+			}
+
+			if pv == nil {
+				return fmt.Errorf("PersistentVolume not found: %q", pvName)
+			}
+
+			if pv.Spec.AWSElasticBlockStore != nil {
+				ebsVolumes[pv.Spec.AWSElasticBlockStore.VolumeID] = true
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *MaxEBSVolumeCountChecker) predicate(pod *api.Pod, existingPods []*api.Pod, node string) (bool, error) {
+	newEBSVolumes := make(map[string]bool)
+	if err := c.filterEBSVolumes(pod.Spec.Volumes, pod.Namespace, newEBSVolumes); err != nil {
+		return false, err
+	}
+
+	// quick return
+	if len(newEBSVolumes) == 0 {
+		return true, nil
+	}
+
+	// count unique volumes
+	existingEBSVolumes := make(map[string]bool)
```

Comment:
From what I saw in the EBS code, if the volume is already mounted, it will use the same device, so it shouldn't count towards the limit.  Someone more intimately familiar with the EBS mounter should feel free to correct me on that, though.

git blame points me to @justinsb, perhaps?